### DISCUSSION TITLE
QoS Enforcement and UE MTU

### DISCUSSION THREAD

**Discussion by infinitydon** [aligungr/UERANSIM#456] [Q&A]
Hi,

Please I will like to know if UERANSIM fully honors QoS values from the AMF?

Also if SMF specifies a certain MTU and this is passed to UERANSIM gNB, how will UERANSIM behave?

**Comment by aligungr** [comment]
Hello @infinitydon

Unfortunately UERANSIM completely ignores the QoS values.

And also our gNodeB does not account a custom MTU size.

Best Regards

**Reply by infinitydon** [reply]
@aligungr - I have an issue with OpenAirInterface that when the QoS value is changed in the SMF it affects the iperf throughput and the SMF/UPF developer claims that QoS enforcement is not done also in the UPF this was confirmed in the session request that was sent to the UPF so the only node that can perform QoS should be the gNB but am surprised now that you said that is not the case..

I am sorry to ask again, are you 100% sure that ueransim completely ignores the QoS values from the AMF?

**Reply by aligungr** [reply]
@infinitydon

Yeah, UERANSIM does not process any of the QoS related procedures, and ignores them completely.

Probably the only thing is AMBR, but it's beyond the QoS.

gNodeB *only* handles session and UE AMBR values. (Aggregated maximum bit rate)

**Reply by infinitydon** [reply]
Ok makes sense now, I was mixing up the QoS with the AMBR..

![image](https://user-images.githubusercontent.com/6318992/145062271-b90c048e-4df9-4594-9599-acc02b3e0e4e.png)

**Reply by infinitydon** [reply]
@aligungr - Please if I want to make the gNB not to use the AMBR, how can I do this? Though I am in discussion with the AMF developer to make the AMBR configurable but I need to proceed with my tests to see the highest throughput I can get..

**Reply by aligungr** [reply]
@infinitydon There are 2 methods controling the AMBR in `gnb/gtp/utils.cpp`, that are `allowDownlinkPacket` and `allowUplinkPacket`

You can replace the content of these functions to `return true` always. For example:

```
bool RateLimiter::allowDownlinkPacket(uint64_t pduSession, uint64_t packetSize)
{
    return true;
}
```

and


```
bool RateLimiter::allowUplinkPacket(uint64_t pduSession, uint64_t packetSize)
{
    return true;
}
```

**Reply by infinitydon** [reply]
Thanks @aligungr - I have changed the file but still not having any effect:

```
bool RateLimiter::allowDownlinkPacket(uint64_t pduSession, uint64_t packetSize)
{

    return true;
}

bool RateLimiter::allowUplinkPacket(uint64_t pduSession, uint64_t packetSize)
{

    return true;
}
```

It is still honoring the AMBR from the AMF

**Reply by aligungr** [reply]
Hmm I think it should work. Please make sure that you successfully built the UERANSIM after the change.

**Reply by infinitydon** [reply]
Yes I built the UERANSIM (v3.2.5) successfully without any errors when I modified the file

**Reply by infinitydon** [reply]
I did some more test by setting the ambr values to low speed in the SMF:

<img width="1054" alt="image" src="https://user-images.githubusercontent.com/6318992/145818949-ce01b21a-f6c5-4cf9-8cfc-6f217e13be9e.png">

As shown in the screenshot, PDU session ambr limit is 20/22Mbps and the iperf shows ~ 1.7 Gbps

```
ubuntu@ip-10-0-3-8:~/ambr_test/UERANSIM-3.2.5/build$ ./nr-binder 12.1.1.2 iperf3 -c 10.0.5.8 -i 1 -t 20
Connecting to host 10.0.5.8, port 5201
[  5] local 12.1.1.2 port 54309 connected to 10.0.5.8 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   230 MBytes  1.93 Gbits/sec   41    288 KBytes
[  5]   1.00-2.00   sec   209 MBytes  1.75 Gbits/sec  137    222 KBytes
[  5]   2.00-3.00   sec   202 MBytes  1.70 Gbits/sec  150    205 KBytes
[  5]   3.00-4.00   sec   196 MBytes  1.65 Gbits/sec  168    276 KBytes
[  5]   4.00-5.00   sec   208 MBytes  1.74 Gbits/sec  180    197 KBytes
[  5]   5.00-6.00   sec   218 MBytes  1.82 Gbits/sec  107    259 KBytes
[  5]   6.00-7.00   sec   206 MBytes  1.73 Gbits/sec   60    304 KBytes
[  5]   7.00-8.00   sec   214 MBytes  1.79 Gbits/sec  109    204 KBytes
[  5]   8.00-9.00   sec   206 MBytes  1.73 Gbits/sec  105    232 KBytes
[  5]   9.00-10.00  sec   205 MBytes  1.72 Gbits/sec  166    199 KBytes
[  5]  10.00-11.00  sec   191 MBytes  1.60 Gbits/sec  243    258 KBytes
[  5]  11.00-12.00  sec   171 MBytes  1.44 Gbits/sec  432    136 KBytes
[  5]  12.00-13.00  sec   178 MBytes  1.49 Gbits/sec  246    195 KBytes
[  5]  13.00-14.00  sec   170 MBytes  1.43 Gbits/sec  360    195 KBytes
[  5]  14.00-15.00  sec   192 MBytes  1.61 Gbits/sec  194    203 KBytes
[  5]  15.00-16.00  sec   208 MBytes  1.74 Gbits/sec  127    209 KBytes
[  5]  16.00-17.00  sec   198 MBytes  1.66 Gbits/sec  186    208 KBytes
[  5]  17.00-18.00  sec   208 MBytes  1.74 Gbits/sec   77    362 KBytes
[  5]  18.00-19.00  sec   216 MBytes  1.81 Gbits/sec  120    228 KBytes
[  5]  19.00-20.00  sec   192 MBytes  1.61 Gbits/sec  135    182 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-20.00  sec  3.92 GBytes  1.68 Gbits/sec  3343             sender
[  5]   0.00-20.00  sec  3.92 GBytes  1.68 Gbits/sec                  receiver
```
_**So I think what you suggested is working (apologies for not testing widely), so now it seems the maximum speed that ueransim can support might be around 2Gbps?**_

I also tried a UERANSIM compiled without AMBR code changes just to see if the speed will be applied to the PDU session and the throughput matches the one that is set in SMF:

```
ubuntu@ip-10-0-3-8:~/UERANSIM-3.2.4/build$ ./nr-binder 12.1.1.4 iperf3 -c 10.0.5.8 -i 1 -t 20
Connecting to host 10.0.5.8, port 5201
[  5] local 12.1.1.4 port 33865 connected to 10.0.5.8 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  7.09 MBytes  59.4 Mbits/sec  692   81.6 KBytes
[  5]   1.00-2.00   sec  2.50 MBytes  21.0 Mbits/sec  342   3.95 KBytes
[  5]   2.00-3.00   sec  2.50 MBytes  21.0 Mbits/sec  280   38.2 KBytes
[  5]   3.00-4.00   sec  2.50 MBytes  21.0 Mbits/sec  278   17.1 KBytes
[  5]   4.00-5.00   sec  2.50 MBytes  21.0 Mbits/sec  316   14.5 KBytes
[  5]   5.00-6.00   sec  2.50 MBytes  21.0 Mbits/sec  202   18.4 KBytes
[  5]   6.00-7.00   sec  1.25 MBytes  10.5 Mbits/sec  203   27.6 KBytes
[  5]   7.00-8.00   sec  2.50 MBytes  21.0 Mbits/sec  239   27.6 KBytes
[  5]   8.00-9.00   sec  2.50 MBytes  21.0 Mbits/sec  309   2.63 KBytes
[  5]   9.00-10.00  sec  2.50 MBytes  21.0 Mbits/sec  242   14.5 KBytes
[  5]  10.00-11.00  sec  2.50 MBytes  21.0 Mbits/sec  309   13.2 KBytes
[  5]  11.00-12.00  sec  1.25 MBytes  10.5 Mbits/sec  201   5.27 KBytes
[  5]  12.00-13.00  sec  3.75 MBytes  31.5 Mbits/sec  316   6.58 KBytes
[  5]  13.00-14.00  sec  1.25 MBytes  10.5 Mbits/sec  174   43.4 KBytes
[  5]  14.00-15.00  sec  2.50 MBytes  21.0 Mbits/sec  243   15.8 KBytes
[  5]  15.00-16.00  sec  2.50 MBytes  21.0 Mbits/sec  255   26.3 KBytes
[  5]  16.00-17.00  sec  2.50 MBytes  21.0 Mbits/sec  224   30.3 KBytes
[  5]  17.00-18.00  sec  2.50 MBytes  21.0 Mbits/sec  282   5.27 KBytes
[  5]  18.00-19.00  sec  1.25 MBytes  10.5 Mbits/sec  205   31.6 KBytes
[  5]  19.00-20.00  sec  2.50 MBytes  21.0 Mbits/sec  285   17.1 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-20.00  sec  50.8 MBytes  21.3 Mbits/sec  5597             sender
[  5]   0.00-20.00  sec  48.1 MBytes  20.2 Mbits/sec                  receiver

iperf Done.
```

I also attached the pcap trace just in case.

[ueransim-dec-13.zip](https://github.com/aligungr/UERANSIM/files/7704078/ueransim-dec-13.zip)

**Reply by infinitydon** [reply]
Am on a system that is capable of pushing above 9Gbps:

```
ubuntu@ip-10-0-3-8:~/ambr_test/UERANSIM-3.2.5$ iperf3 -c 10.0.5.8 -i 1 -t 5
Connecting to host 10.0.5.8, port 5201
[  5] local 10.0.3.8 port 45768 connected to 10.0.5.8 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  1.11 GBytes  9.56 Gbits/sec    0   1.61 MBytes
[  5]   1.00-2.00   sec  1.11 GBytes  9.53 Gbits/sec    0   1.69 MBytes
[  5]   2.00-3.00   sec  1.11 GBytes  9.53 Gbits/sec    0   1.69 MBytes
[  5]   3.00-4.00   sec  1.11 GBytes  9.53 Gbits/sec    0   1.69 MBytes
[  5]   4.00-5.00   sec  1.11 GBytes  9.53 Gbits/sec    0   1.69 MBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-5.00   sec  5.55 GBytes  9.54 Gbits/sec    0             sender
[  5]   0.00-5.00   sec  5.55 GBytes  9.53 Gbits/sec                  receiver

iperf Done.
```

**Reply by aligungr** [reply]
> So now it seems the maximum speed that ueransim can support might be around 2Gbps?

Actually there is no theoretical limit for the maximum throughput of UERANSIM. But, since the IP packets are processed by the CPU inside of the UE and gNodeB application, hardware (especially CPU) directly influences the throughput of the UERANSIM. Therefore you can get lower or higher than this throughput on different systems. 

(However I'm pretty sure that we can increase the throughput by optimizing the gNodeB, but currently I'm unable to allocate some time for this purpose.)
